feat: Add whatsapp template builder

Aman-14 opened the pull request:
# Pull Request

## Description

The PR includes a template manager for whatsapp to create templates from Chatwoot #10886

Feature Overview

  - Adds full WhatsApp message template management: REST endpoints for list/show/
  create/delete ([app/controllers/api/v1/accounts/message_templates_controller.rb](https://github.com/chatwoot/chatwoot/pull/12231/files#diff-651266cc61affd0a16328a03edc0b0f113abbd07d8029ad47af57fb6a85825b6R1)),
  new MessageTemplate model with policy and migration ([db/migrate/20250901081601_create_message_templates.rb](https://github.com/chatwoot/chatwoot/pull/12231/files#diff-baabcfb95cf3c13ffb02c8169ac0d0f5d33585a9d78c5ce033d2ab3cf1273317R1)),
  plus formatter/sync services for Meta integration ([app/services/whatsapp/
  template_formatter_service.rb](https://github.com/chatwoot/chatwoot/pull/12231/files#diff-4761b267f4631aa8d70d7e489842112e0773b8b43950c07d733645d724b30dd3R2), [app/services/whatsapp/template_sync_service.rb](https://github.com/chatwoot/chatwoot/pull/12231/files#diff-ecc2cf7105a702f5eb68e44be17d1142b4358c933796cc936fb4324095059828R1)).
  - Extends WhatsApp Cloud provider flow to handle template creation with media
  uploads by introducing a dedicated Facebook upload client and restructuring
  creation logic ([app/services/whatsapp/template_creation_service.rb](https://github.com/chatwoot/chatwoot/pull/12231/files#diff-e8788a61a62cfeaec8191feb1483df36b246c01f77a416758b15bd31c7353826R1), [app/
  services/whatsapp/facebook_upload_service.rb](https://github.com/chatwoot/chatwoot/pull/12231/files#diff-ed84ba7056d8fb5290b6163c10056771d399671fd80c55302e1dcef78be66f3aR2), [app/services/whatsapp/providers/
  whatsapp_cloud_service.rb](https://github.com/chatwoot/chatwoot/pull/12231/files#diff-1c1b4c147a74aeb2d0abe69da6788ee4410382290e52c2d8b59b70685110b7a2R201))
- This feature is gated behind a Template builder feature flag

Frontend

  - Introduces template builder UI with modular header/body/footer/buttons sections,
  routing, and Vuex store to manage templates ([app/javascript/dashboard/routes/
  dashboard/settings/templates/TemplateBuilderPage.vue](https://github.com/chatwoot/chatwoot/pull/12231/files#diff-59e150c626d395261e3ce6afd787a2c8297f227ada6bcf252ef4af54bbae1469R1), [app/javascript/dashboard/
  store/modules/messageTemplates.js](https://github.com/chatwoot/chatwoot/pull/12231/files#diff-a661c6eead09b4c605d23b6455f38af4da0841828ad2a1970341d78d9df88a30R1)).

Media Uploads
- Media upload happens [FileUpload](https://github.com/chatwoot/chatwoot/pull/12231/files#diff-51dcad97957ec41805aba0a9226e7db1cb6555852e6a9551caf49014ab8aadd3R1) component that uses `/api/v1/accounts/:accountId/upload` API to upload files, which then uploaded to Meta using `facebook_upload_service` to meta before template creation. 
- I am not doing any cleanup of uploaded files. 

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

I have tested it manually and also added specs.


## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented on my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

Aman-14 committed:
feat(whatsapp): Add create_message_template method for WhatsApp message templates

Aman-14 commented:
@sojan-official where should we put this template manager on frontend? We can add it under inboxes because these templates will be inbox specific. The image in #10886 shows it under settings tab in sidebar, but each WhatsApp inbox has its own waba and templates belong to the waba, not the Chatwoot account. This means we can't show "all templates" globally since they're scoped per business account. Please correct me if I am wrong.

muhsin-k commented:
> @sojan-official where should we put this template manager on frontend? We can add it under inboxes because these templates will be inbox specific. The image in #10886 shows it under settings tab in sidebar, but each WhatsApp inbox has its own waba and templates belong to the waba, not the Chatwoot account. This means we can't show "all templates" globally since they're scoped per business account. Please correct me if I am wrong.

@Aman-14 We are considering implementing this at the account level rather than the inbox level. This approach would require moving all message templates to a separate model to unify across all levels.

Aman-14 committed:
feat: add templates page

Aman-14 committed:
add todo to fix eslint issue

Aman-14 committed:
Merge branch 'develop' into feature/template-manager

Aman-14 committed:
feat: update template builder sections

Aman-14 committed:
Merge branch 'develop' into feature/template-manager

Aman-14 committed:
feat: add message template schema

Aman-14 commented:
@sojan-official @muhsin-k This is the template schema I have created, please share your thoughts

```
# Table name: message_templates
#
#  id                   :bigint           not null, primary key
#  category             :integer          default(0), not null
#  channel_type         :string(50)       not null
#  content              :jsonb            not null
#  language             :string(10)       default("en"), not null
#  last_synced_at       :datetime
#  metadata             :jsonb
#  name                 :string(255)      not null
#  status               :integer          default(0)
#  created_at           :datetime         not null
#  updated_at           :datetime         not null
#  account_id           :bigint           not null
#  created_by_id        :bigint
#  inbox_id             :bigint
#  platform_template_id :string(255)
#  updated_by_id        :bigint
#
# Indexes
#
#  idx_on_account_id_name_language_channel_type_99a8ab867f  (account_id,name,language,channel_type) UNIQUE
#  index_message_templates_on_account_id                    (account_id)
#  index_message_templates_on_channel_type                  (channel_type)
#  index_message_templates_on_created_by_id                 (created_by_id)
#  index_message_templates_on_inbox_id                      (inbox_id)
#  index_message_templates_on_status                        (status)
#  index_message_templates_on_updated_by_id                 (updated_by_id)

```

Aman-14 committed:
feat: add template controller

- add template controller with very basic validation
- add policy for authorizatoin
- add jbuilder jsonfiles
- update routes.rb to include message template routes

muhsin-k commented:
> Hi @sojan-official @muhsin-k This is the template schema I have created, please share your thoughts
> 
> ```
> # Table name: message_templates
> #
> #  id                   :bigint           not null, primary key
> #  category             :integer          default(0), not null
> #  channel_type         :string(50)       not null
> #  content              :jsonb            not null
> #  language             :string(10)       default("en"), not null
> #  last_synced_at       :datetime
> #  metadata             :jsonb
> #  name                 :string(255)      not null
> #  status               :integer          default(0)
> #  created_at           :datetime         not null
> #  updated_at           :datetime         not null
> #  account_id           :bigint           not null
> #  created_by_id        :bigint
> #  inbox_id             :bigint
> #  platform_template_id :string(255)
> #  updated_by_id        :bigint
> #
> # Indexes
> #
> #  idx_on_account_id_name_language_channel_type_99a8ab867f  (account_id,name,language,channel_type) UNIQUE
> #  index_message_templates_on_account_id                    (account_id)
> #  index_message_templates_on_channel_type                  (channel_type)
> #  index_message_templates_on_created_by_id                 (created_by_id)
> #  index_message_templates_on_inbox_id                      (inbox_id)
> #  index_message_templates_on_status                        (status)
> #  index_message_templates_on_updated_by_id                 (updated_by_id)
> ```

@Aman-14 Have you considerd the [twilio content templates](https://github.com/chatwoot/chatwoot/pull/12272) as well?

muhsin-k commented:
> > 
> > @Aman-14 Have you considerd the [twilio content templates](https://github.com/chatwoot/chatwoot/pull/12272) as well?
> 
> @muhsin-k I didn't check this PR yet, but I have made the schema generic so it should support other template structure. The `content` field is of jsonb type -- we should be able to save any template structure.

@Aman-14 Got it. Please keep in mind that we might need to migrate the existing templates to the new structure

Aman-14 committed:
feat: add template sync with meta

- Add validations for inbox, category, and channel_type in
MessageTemplate model
- Implement template creation on Meta platform before local save
- Create TemplateSyncService to sync templates from Meta to
MessageTemplate model
- Add proper error handling for Meta API responses with
user-friendly messages
- Remove unnecessary inbox fetching in controller
- Ensure templates synced from Meta don't trigger recreation via
platform_template_id check

Aman-14 committed:
chore: fix robocop issue

Aman-14 committed:
feat: improve builder UI with reusable components

Aman-14 commented:
Hi @muhsin-k  I am working on the media upload part of template manager. I am using `/api/v1/accounts/{id}/upload` endpoint to directly upload the media to backend, and when user submits the template then I am planning to fetch the media from ActiveStorage and upload it to meta, but when users upload media files through this endpoint there's currently no way to clean them up if the user abandons the process or removes the file.
This means uploaded files just sit in ActiveStorage forever, which could lead to storage bloat over time. I can simply add a delete endpoint here, but wanted to check if this approach aligns with how you'd want to handle blob cleanup?

Aman-14 committed:
feat: add frontend media support in template manager

- Add FileUpload component for IMAGE/VIDEO/DOCUMENT handling
- Update HeaderSection with FileUpload integration
- Refactor TemplateBuilder to use imperative API pattern

Aman-14 committed:
Merge develop

Aman-14 committed:
feat: add media upload service

Implements https://developers.facebook.com/docs/graph-api/guides/upload

Aman-14 committed:
fix: fix component generation

- Add conditional example generation to avoid empty example objects
- Fix header format icons by moving definitions to component scope
- Use BUTTONS type for buttons
- Disable Authentication category with "Coming Soon" indicator
- Add 10-button limit with disabled state for add button

Aman-14 committed:
refactor: extract WhatsApp template creation to dedicated service

- Create Whatsapp::TemplateCreationService that handles whatsapp
  template creation
- Move template validation and creation logic out of MessageTemplate
model
- Add media upload integration with FacebookApiClient for template assets

Aman-14 commented:
> > > @Aman-14 Have you considerd the [twilio content templates](https://github.com/chatwoot/chatwoot/pull/12272) as well?
> > 
> > 
> > @muhsin-k I didn't check this PR yet, but I have made the schema generic so it should support other template structure. The `content` field is of jsonb type -- we should be able to save any template structure.
> 
> @Aman-14 Got it. Please keep in mind that we might need to migrate the existing templates to the new structure.

Hi @muhsin-k ,

I've been thinking about the migration and thought of a lazy migration approach using method override instead of upfront migration.

Override the existing `message_templates` JSONB field with a method in `Channel::Whatsapp` that checks for legacy data when accessed, migrates it automatically to the new `MessageTemplate` table using
`TemplateSyncService`, and returns data from the database in backward-compatible format.

```ruby
def message_templates
  legacy_message_templates = read_attribute(:message_templates)

  if legacy_message_templates.present? && legacy_message_templates.length > 0
    sync_service = Whatsapp::TemplateSyncService.new(
      channel: self,
      templates: legacy_message_templates
    )
    if sync_service.call
      # Clear legacy data after sync
      self[:message_templates] = nil
      save!
    end
  end

  db_templates = account.message_templates
                        .by_channel_type('Channel::Whatsapp')
                        .by_inbox(inbox.id)
  db_templates.map do |template|
    {
      'id' => template.id.to_s,
      'name' => template.name,
      'language' => template.language,
      'status' => template.status,
      'category' => template.category,
      'parameter_format' => template.parameter_format,
      'components' => template.content['components'] || []
    }  
  end
end
```

This won't break any existing usage of channel.message_templates and will give us enough time to update if needed.

What are your thoughts?

Aman-14 committed:
chore: clean duplicate constants

Aman-14 committed:
fix: bug fixes
- use assign_attributes instead of update! in
  message_templates_controller

Aman-14 committed:
chore: rename MediaUploadService to FacebookUploadService
- This is to potentially support media upload endpoint
of whatsapp cloud service
https://developers.facebook.com/docs/whatsapp/cloud-api/reference/media#upload-media

Aman-14 committed:
tests: add tests for backend code

Aman-14 committed:
Merge develop

Aman-14 committed:
feat: add template_builder feature flag and restrict WhatsApp template sync

- Add template_builder feature flag configuration in features.yml
- Add TEMPLATE_BUILDER constant to frontend featureFlags.js
- Restrict templates routes to admin-only access with feature flag protection
- Modify WhatsApp Cloud Service to only call TemplateSyncService when feature is enabled
- Add tests for both enabled and disabled feature flag scenarios
- Remove redirect from templates root path - list now shows directly on
/settings/templates
- Update route permissions from all roles to administrator-only

Aman-14 committed:
chore: parse variables again when parameter format is changed

Aman-14 committed:
chore(templates): move helpers in helper file

Aman-14 committed:
fix(helper): resolve regex reuse issues and improve variable parsing

- Convert regex exports to factory functions to prevent global state issues with .test
- Fix variable count logic in parsePositionalVariables with Math.min capping
- Add early error return in parseNamedVariables to avoid unnecessary processing
- Remove unused extractPositionalVariables function

Aman-14 committed:
tests: add tests for templateHelper

Aman-14 committed:
chore: remove edit support

Aman-14 committed:
chore: remove :update from backend

Aman-14 committed:
feat: add delete modal and sync templates

Aman-14 committed:
feat: add support for button url dynamic variable

Aman-14 committed:
Merge branch 'develop' into feature/template-manager

Aman-14 committed:
tests: update templateHelper tests
- add tests for error property in button
- add tests for validateTemplateData

Aman-14 committed:
Merge branch 'develop' into feature/template-manager

Aman-14 committed:
fix: reset template builder state on deactive

Aman-14 committed:
feat(tests): add tests for Message template api client and store

Aman-14 committed:
fix: do in-memory filteration instead of fetching again

Aman-14 committed:
Merge branch 'develop' into feature/template-manager

Aman-14 committed:
fix: send full message_templates params

Aman-14 committed:
chore: cleanup

Aman-14 commented:
Hi @muhsin-k @sojan-official , this is what I have built so far. 

https://github.com/user-attachments/assets/2998aaf5-2376-45e7-b3ed-c6831591be7a

Please take a look at the PR and let me know if you want me to add/update anything. I have update the PR description and added summary of changes

sojan-official commented:
@Aman-14 looks pretty dope 🔥  .  @muhsin-k  will take a pass and suggest the next steps  !!!

Aman-14 committed:
Merge branch 'develop' into feature/template-manager

Aman-14 committed:
Merge branch 'develop' into feature/template-manager

Aman-14 committed:
feat: add category based grouping of buttons and constants
- categories: cta and quickReply

Aman-14 committed:
feat: update template builder preview and shared button helpers

Aman-14 committed:
Merge branch 'develop' into feature/template-manager

Aman-14 committed:
fix: avoid re-triggering provider sync during template imports

stenioanibal commented:
@sojan-official @muhsin-k any change to get this on the next release?

Aman-14 committed:
fix(message_templates): accept signed blob id for template media upload

Aman-14 committed:
fix(message_templates): enqueue WhatsApp template sync job after create

Aman-14 committed:
feat(message_templates): add delay to creation hook

Aman-14 committed:
Merge develop

propiter commented:
Hi @Aman-14 👋 — I've been testing this PR locally and found two bugs that cause 3 of the 11 specs in `template_creation_service_spec.rb` to fail. Sharing in case it helps move this forward.

---

### Bug 1 — `create_message_template` and `delete_message_template` are private

Both methods are defined **after** the `private` keyword in `WhatsappCloudService` (around line 218), but they are called from outside the class:

- `TemplateCreationService` calls `provider_service.create_message_template(...)`
- `MessageTemplate` model hooks call `delete_message_template` via the provider

This causes a `NoMethodError: private method called` at runtime. It also makes the RSpec `instance_double` stubs silently fail in tests, since `instance_double` only verifies the public interface.

**Fix:** move both methods to the public section (before `private`), keeping `handle_template_error` private.

---

### Bug 2 — `blob_id` key mismatch (snake_case vs camelCase)

In `TemplateCreationService#process_media_uploads`:

```ruby
# Current (broken) — key is never found, blob_id is always nil
blob_id = header_component['media']['blob_id']

# Fix — matches the camelCase key the frontend actually stores
blob_id = header_component['media']['blobId']
```

The frontend stores the key as `blobId` (camelCase), but the service reads `blob_id` (snake_case) → always `nil` → media uploads silently fail with no error surfaced to the user.

---

After both fixes, all **92 specs** in the template builder suite pass green. Happy to open a PR to your branch if that helps!

propiter commented:
Hi @Aman-14 — follow-up to my previous comment.

While continuing to test this PR rebased on current `develop`, I hit a separate but quite serious issue. Sharing here because it directly affects how this PR should land.

### The problem

This PR inserts `template_builder` between `saml` and `advanced_search_indexing` in `config/features.yml` (around line 207). Because positions in that file map to bit positions in the `accounts.feature_flags` bigint bitmask (via `flag_shih_tzu`), inserting in the middle shifts every subsequent feature by one bit.

Once rebased on `develop` (which has `advanced_assignment` from #13036 at the tail), the shift pushes `advanced_assignment` to **bit 63** — the sign bit of a signed `bigint`. Any account whose bitmask ends up with that bit set raises:

```
ActiveModel::RangeError (18444351259918991351 is out of range
                        for ActiveModel::Type::Integer with limit 8 bytes)
  enterprise/app/controllers/enterprise/super_admin/accounts_controller.rb:24
```

In our environment, after deploying this PR, **12 accounts** ended up with silently misaligned feature lists (`advanced_search_indexing` reported as `template_builder`, `reply_mailer_migration` as `advanced_search_indexing`, and so on for the 9 features in between). For accounts whose shift set bit 63, every SuperAdmin save crashed with the `RangeError` above.

This is exactly what the header comment in the file itself warns against:

```yaml
# DO NOT change the order of features EVER
```

### Suggested fix

Move the new feature to the very end of the file:

```diff
--- a/config/features.yml
+++ b/config/features.yml
@@ -204,9 +204,6 @@
   display_name: SAML
   enabled: false
   premium: true
-- name: template_builder
-  display_name: Template Builder
-  enabled: false
 - name: advanced_search_indexing
   display_name: Advanced Search Indexing
@@ -243,3 +240,6 @@
   display_name: Advanced Assignment
   enabled: false
   premium: true
+- name: template_builder
+  display_name: Template Builder
+  enabled: false
```

In our fork this restored bit positions for all 12 affected accounts **without any database migration** — the stored bitmask values were already correct under the original ordering; only the interpretation had drifted.

### A broader concern worth flagging

Regardless of where `template_builder` lands here, the ceiling is close: `develop` is already at 63 features after `advanced_assignment`, and `accounts.feature_flags` is a signed `bigint` (max `2^63 - 1`). The next feature added — whichever it is — will land on bit 63 and trigger the same `RangeError` for any account that enables it.

I'll open a separate issue to discuss that broader problem so it doesn't derail this thread, along with a small PR that adds a CI guard against future middle-insertions and accidental count overflow.

Hope this is useful — happy to help move things forward.
